core/src/Kokkos_Parallel_Reduce.hpp

```diff
@@ -109,6 +109,9 @@ struct Sum {
   bool references_scalar() const { return references_scalar_v; }
 };
 
+template <typename Scalar, typename Space>
+Sum(View<Scalar, Space> const&) -> Sum<Scalar, Space>;
```

Comment:
This can't possible work, e.g. I hand it `Kokkos::View<double**, Kokkos::LayoutRight>` ...

Reply:
This seems like an orthogonal issue.  The `Space` argument in the reducer `Sum<Scalar, Space>` is already unconstrained.  I don't think it should be solved as part of adding CTADs.

We should probably open a separate issue for this.  My first thought is that the fix for all the reducers is to replace `Space` with `Properties...` to match the template parameters for `View`.

Reply:
It is not an orthogonal issue. 

I can call the constructor of

```
Sum<int, HostSpace> 
```

With all of the following:
```c++
View<int, HostSpace>
View<int, LayoutRight, HostSpace>
View<int, LayoutLeft, HostSpace>
View<int, LayoutStride, HostSpace, MemoryTraits<Unmanaged>>
...
```

But the deduction would not deduce as written. 
I want to store a plain View, no memory traits no layout (since its scalar, which we may need to fix in the future), no other extraneous information. 
And I want to still have the thing deduced correct. 

So I think we need this:
```c++
template <typename Scalar, class ... Properties>
Sum(View<Scalar, Properties ... > const&) -> Sum<Scalar, typename View<Scalar, Properties ... >::memory_space>;
```

Reply:
In the end we can write deduction guides where the supposed argument of the constructor is anything convertible to the actual constructor argument.

Reply:
Oh, I see.  Will try it out.

Reply:
I don't think we can quite do anything implicitly convertible to `View<Scalar, Space>` because I have to get the `memory_space` out of the type of the argument passed in, which we cannot do for an arbitrary type.

We can, however, do any `View<Scalar, Properties...>`, which should be sufficient (assuming those are implicitly convertible to `View<Scalar, View<Scalar, Properties...>::memory_space>`, although I could check that too).

In the deduction guides, should I constrain `Scalar` to be a scalar (and by "scalar" I mean non-pointer non-array, or is it something else)?

Reply:
I think the sensible phrasing of that latter check is that the View has rank 0?

Reply:
And is not `const`.

The question is:  where does the scalar check belong?  IMO, they belong in the reducers, not the deduction guides.

Also, putting SFINAE checks in a deduction guide can make error messages worse, not better, depending on the compiler.